docs-client/src/components/VariableList/index.tsx

```diff
@@ -83,168 +90,136 @@ export default function({
-                </TableCell>
-                <TableCell>
-                  <code>
-                    {this.props.specification.getTypeSignatureHtml(
-                      variable.typeSignature,
-                    )}
-                  </code>
-                </TableCell>
-                <TableCell>{variable.docString}</TableCell>
-                {this.props.hasBean && this.showExpandIfBean(index)}
-              </TableRow>
-              {this.state.openBeans[index] && (
-                <FieldInfos
-                  {...this.props}
-                  indent={this.props.indent + 2}
-                  variables={variable.childFieldInfos!}
-                  state={this.state.childStates[index]}
-                  childIndex={index}
-                  setChildStateInParent={this.setChildStateInParent}
-                />
-              )}
-            </React.Fragment>
-          ))
-        ) : (
-          <TableRow>
-            <TableCell colSpan={this.state.colSpanLength}>
-              There are no {this.props.title.toLowerCase()}
-            </TableCell>
-          </TableRow>
-        )}
-      </>
-    );
-  }
+  const [expanded, toggleExpanded] = useReducer((value) => !value, false);
 
-  private setChildStateInParent(index: number, childState: State) {
-    this.state.childStates[index] = childState;
-  }
+  const hasChildren =
+    variable.childFieldInfos && variable.childFieldInfos.length > 0;
 
-  private showExpandIfBean(index: number) {
-    return (
-      <>
+  return (
+    <>
+      <TableRow
+        onClick={toggleExpanded}
+        className={hidden ? styles.hidden : ''}
+      >
         <TableCell>
-          {this.state.isBeans[index] &&
-            (this.state.openBeans[index] ? <ExpandLess /> : <ExpandMore />)}
+          <code>{indentString(indent, variable.name)}</code>
         </TableCell>
-      </>
-    );
+        {hasLocation && variable.location && (
+          <TableCell>
+            <code>{formatLocation(variable.location)}</code>
+          </TableCell>
+        )}
+        <TableCell>
+          <code>{formatRequirement(variable.requirement)}</code>
+        </TableCell>
+        <TableCell>
+          <code>
+            {specification.getTypeSignatureHtml(variable.typeSignature)}
+          </code>
+        </TableCell>
+        <TableCell>{variable.docString}</TableCell>
+        {hasChildren && (
+          <TableCell>{expanded ? <ExpandLess /> : <ExpandMore />}</TableCell>
+        )}
+      </TableRow>
+      {hasChildren && (
+        <FieldInfos
+          hasLocation={hasLocation}
+          hidden={hidden || !expanded}
+          indent={indent + 2}
+          variables={variable.childFieldInfos!}
+          specification={specification}
+          title={title}
+        />
+      )}
+    </>
+  );
+};
+
+const FieldInfos: React.FunctionComponent<FieldInfosProps> = (props) => {
+  const styles = useStyles();
+
+  const isEmpty = props.variables.length === 0;
```

Comment:
Question: Here, we are assuming the variable list won't change in runtime, right?

Reply:
Not really - if the specification is refetched and something changes, all the components will rerender to reflect the change when props change.

Reply:
Got it. Thanks!